---
title: 'Loading SRA, VAST-TOOLS and user-provided RNA-seq data'
author: "Nuno Saraiva-Agostinho"
date: "`r Sys.Date()`"
bibliography: refs.bib
csl: bioinformatics.csl
output: 
    rmarkdown::html_vignette:
        toc: true
vignette: >
  %\VignetteIndexEntry{Loading SRA, VAST-TOOLS and user-provided RNA-seq data}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

---

*psichomics* is an interactive R package for integrative analyses of alternative
splicing and gene expression based on [The Cancer Genome Atlas (TCGA)][tcga]
(containing molecular data associated with 34 tumour types), the
[Genotype-Tissue Expression (GTEx)][gtex] project (containing data for multiple
normal human tissues), [Sequence Read Archive][SRA] and user-provided data.

# Preparing generic data

psichomics now supports loading alternative splicing quantification and gene
expression output directly from VAST-TOOLS
([read here how](#loading-data-from-vast-tools)). psichomics also supports
importing generic data from any source as long as the tables are prepared as
detailed below.

> Please make sure that sample and subject identifiers are exactly identical
between datasets.

## Sample information

- Tab-separated values (TSV)
- **Sample identifiers (rows)** and their **attributes (columns)**
- The first column must contain sample identifiers and be named `Sample ID`
- Optionally, indicate the subject associated to each sample in a column named
`Subject ID`

> Sample information from the SRA Run Selector can be automatically loaded using
the visual interface or [`loadLocalFiles()`][loadLocalFiles]

- Example of a valid sample information dataset:

| Sample ID | Type   | Tissue | Subject ID |
| --------- | ------ | ------ | ---------- |
| SMP-01    | Tumour | Lung   | SUBJ-03    |
| SMP-02    | Normal | Blood  | SUBJ-12    |
| SMP-03    | Normal | Blood  | SUBJ-25    |
    
## Subject information

- Tab-separated values (TSV)
- **Subject identifiers (rows)** and their **attributes (columns)**
- The first column must contain subject identifiers and be named `Subject ID`
- Example of a valid subject information dataset:

| Subject ID | Age | Gender | Race  |
| ---------- | ---:| ------ | ----- |
| SUBJ-01    |  34 | Female | Black |
| SUBJ-02    |  22 | Male   | Black |
| SUBJ-03    |  58 | Female | Asian |

## Gene expression

- Tab-separated values (TSV)
- Read counts of **genes (rows)** across **sample (columns)**
- The first column must contain gene symbols and be named `Gene ID`
- Example of a valid gene expression dataset:
    
| Gene ID | SMP-18 | SMP-03 | SMP-54 |
| ------- | ------:| ------:| ------:|
| AMP1    |     24 |     10 |     43 |
| BRCA1   |     38 |     46 |     32 |
| BRCA2   |     43 |     65 |     21 |
    
## Exon-exon junction quantification

- Tab-separated values (TSV)
- Read counts of **exon-exon junctions (rows)** across **samples (columns)**
- The first column must contain junction identifiers and be named `Junction ID`
- Only chromosome number and capital letters X, Y, Z, W, and M are supported, 
followed by the genomic regions; acceptable junction identifiers include:
    - `10_18748_21822`
    - `chromosome 10 (18748 to 21822)`
    - `chr10:18748-21822`
- Optionally, indicate the strand with `+` or `-` at the end of the junction
identifier:
    - `10:3213:9402:+`
    - `chr10:3213-9402 -`
- Rows whose junction identifiers contain `alt`, `random` or `Un` in chromosome
names are discarded
- Example of a valid exon-exon junction quantification dataset:
    
| Junction ID    | SMP-18 | SMP-03 |
| -------------- | ------:| ------:|
| 10:6752-7393   |      4 |      0 |
| 10:18748-21822 |      8 |     46 |
| 10:24257-25325 |     83 |     65 |
    
## Alternative splicing quantification (also known as inclusion levels)

- Tab-separated values (TSV)
- Alternative splicing quantification values of
**alternative splicing events (rows)** across **samples (columns)**
- The first column must be named `AS Event ID`

> Note that psichomics cannot currently parse alternative splicing events (e.g.
cognate gene and coordinates) from generic PSI tables.

- Values between 0 and 1 or between 0 and 100: if the latter, values are
automatically scaled between 0 and 1
- Example of a valid alternative splicing quantification dataset:
    
| AS Event ID       | SMP-18 | SMP-03 |
| ----------------- | ------:| ------:|
| someASevent001    |   0.71 |   0.30 |
| anotherASevent653 |   0.63 |   0.37 |
| yeatAnother097    |   0.38 |   0.62 |

To load the data, move the files to a new folder and follow the instructions in
[Load user-provided data into psichomics][loading].

# Preparing data based on RNA-seq data

The following section goes through the steps required to load data based on
RNA-seq data:

1. Retrieve FASTQ files and sample-associated information from the 
[Sequence Read Archive (SRA)][SRA] (optional if you already have FASTQ files);
2. Map RNA-seq reads from the FASTQ files against a genome of reference using a
splice-aware aligner ([STAR][] in this case);
3. Merge and prepare the output of such aligners to be correctly interpreted
by psichomics;
4. Load data into psichomics.

## Download data from SRA (optional)

[SRA][] is a repository of biological sequence data that stores data from many
published articles. [SRA][] data may be useful to answer pressing biological
questions using publicly available data.

> The latest versions of psichomics support 
**automatic downloading of SRA data** from [recount2][recount], a resource of
pre-processed data for thousands of SRA projects (including gene read counts, 
splice junction quantification and sample metadata). Check first if the project 
of your interest is not available through this resource, thus making it easier 
to analyse gene expression and alternative splicing for your samples of 
interest.

Data from SRA can be downloaded using the [fasterq-dump][fasterq-dump] command
from [sra-tools][sra-tools]. For instance, to retrieve samples from the 
[SRP126561][SRP126561] project:

```
# List SRA samples
samples=(SRR6368612 SRR6368613 SRR6368614 SRR6368615 SRR6368616 SRR6368617)

# Download samples
fasterq-dump --split-3 ${samples}
```

**`--split-3`**: allows to output one or two FASTQ files for single-end or
paired-end sequencing, respectively (a third FASTQ file may also be returned 
containing orphaned single-end reads obtained from paired-end sequencing data)

Sample-associated data is also available from the [Run Selector][SRP126561]
page. Click **RunInfo Table** to download the whole metadata table for all
samples (usually downloaded in a file named `SraRunTable.txt`).

## Align RNA-seq data to quantify splice junctions

The quantification of each alternative splicing event is based on the proportion
of junction reads that support the inclusion isoform, known as percent 
spliced-in or PSI [@wang2008].

To estimate this value for each splicing event, both alternative splicing
annotation and quantification of RNA-Seq reads aligning to splice junctions 
(junction quantification) are required. While alternative splicing annotation is
provided by the package, junction quantification will need to be prepared from
user-provided data by aligning the RNA-seq reads from FASTQ files to a genome of
reference. As junction reads are required to quantify alternative splicing, a
splice-aware aligner will be used. psichomics currently supports [STAR][]
output.

### [STAR][]

#### Genome index

Before aligning FASTQ samples against a genome of reference, a genome index 
needs to be prepared. 

Start by downloading a FASTA file of the whole genome and a GTF file with
annotated transcript. To run the following example, download the 
[human FASTA and GTF files (hg19 assembly)][hg19-data].

```
mkdir hg19_STAR
STAR --runThreadN 4 \
     --runMode genomeGenerate \
     --genomeDir hg19_STAR \
     --genomeFastaFiles /path/to/hg19.fa \
     --sjdbGTFfile /path/to/hg19.gtf
```

Arguments used in the previous command:

- **`--runThreadN 4`**: run STAR in parallel using 4 threads
- **`--runMode genomeGenerate`**: generate the genome index
- **`--genomeDir`**: directory of the genome index directory (output)
- **`--genomeFastaFiles`**: path to genome file (FASTA format)
- **`--sjdbGTFfile`**: path to transcript annotation (GTF format)

#### Aligning against genome index

After the genome index is generated, the sequences in the FASTQ files are
aligned against the annotated gene and splice junctions from the previously
prepared reference. The following commands make STAR output both gene and
junction read counts into files ending in `ReadsPerGene.out.tab` and
`SJ.out.tab`, respectively.

```
align () {
    echo "Aligning ${1} using STAR..."
    STAR --runThreadN 16 \
         --genomeDir hg19_STAR \
         --quantMode GeneCounts \
         --readFilesCommand zcat \
         --outFileNamePrefix ${1} \
         --readFilesIn ${1}_1.fastq ${1}_2.fastq
}

for each in ${samples}; do
    align "${each}"
done
```

Arguments used in the previous command:

- **`--runThreadN 4`**: run STAR in parallel using 4 threads
- **`--genomeDir`**: directory of the genome index directory (input)
- **`--quantMode GeneCounts`**: count reads aligning per gene
- **`--readFilesCommand`**: command to extract compressed FASTQ files
- **`--outFileNamePrefix`**: output prefix
- **`--readFilesIn`**: FASTQ files to align

## Prepare output for psichomics

To process the resulting data files, open an R console or RStudio and type:

```{r, eval=FALSE}
# Change working directory to where the STAR output is
setwd("/path/to/aligned/output/")

library(psichomics)
prepareGeneQuant(
    "SRR6368612ReadsPerGene.out.tab", "SRR6368613ReadsPerGene.out.tab",
    "SRR6368614ReadsPerGene.out.tab", "SRR6368615ReadsPerGene.out.tab",
    "SRR6368616ReadsPerGene.out.tab", "SRR6368617ReadsPerGene.out.tab")
prepareJunctionQuant("SRR6368612SJ.out.tab", "SRR6368613SJ.out.tab", 
                     "SRR6368614SJ.out.tab", "SRR6368615SJ.out.tab",
                     "SRR6368616SJ.out.tab", "SRR6368617SJ.out.tab")
```

To load the data, move the files (including the SRA metadata) to a new folder
and follow the instructions in
[Load user-provided data into psichomics][loading].

# Preparing VAST-TOOLS data

psichomics supports loading inclusion levels and gene expression tables from
VAST-TOOLS (the tables available after running `vast-tools combine`).

Any sample and/or subject information may also be useful to load. Unless the
sample metadata comes from SRA Run Selector, please ensure that it is recognised
by psichomics, as according to [Preparing generic data][generic].

To load the data, move all files to a new folder (VAST-TOOLS PSI and gene
expression tables and sample/subject information) and follow the instructions in
[Load user-provided data into psichomics][loading].

# Load user-provided data into psichomics

## Load using the visual interface

Start psichomics with the following commands in an R console or RStudio:

```{r, eval=FALSE}
library(psichomics)
psichomics()
```

Then, click **Load user files**. Click the **Folder input** tab and select the
appropriate folder. Finally, click **Load files** to automatically scan and load
all supported files from that folder.

## Load using the command-line interface (CLI)

Use function [`loadLocalFiles()`][loadLocalFiles] with the folder path as an
argument:

```{r, eval=FALSE}
library(psichomics)
data <- loadLocalFiles("/path/to/psichomics/input")
geneExpr      <- data[[1]]$`Gene expression`
junctionQuant <- data[[1]]$`Junction quantification`
sampleInfo    <- data[[1]]$`Sample metadata`
```

# Feedback

All feedback on the program, documentation and associated material (including
this tutorial) is welcome. Please send any comments and questions to:

> Nuno Saraiva-Agostinho (nunoagostinho@medicina.ulisboa.pt)
>
> [Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)][distrans]

# References

[tcga]: https://tcga-data.nci.nih.gov/docs/publications/tcga
[gtex]: https://www.gtexportal.org
[SRA]: https://www.ncbi.nlm.nih.gov/sra
[STAR]: https://github.com/alexdobin/STAR
[distrans]: http://imm.medicina.ulisboa.pt/group/distrans/
[fasterq-dump]: https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump
[sra-tools]: https://github.com/ncbi/sra-tools
[hg19-data]: https://grch37.ensembl.org/info/data/ftp/
[SRP126561]: https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP126561
[recount]: https://jhubiostatistics.shinyapps.io/recount/
[loadLocalFiles]: https://nuno-agostinho.github.io/psichomics/reference/loadLocalFiles
[generic]: #preparing-generic-data
[loading]: #load-user-provided-data-into-psichomics
