<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading SRA and user-provided RNA-seq data â€¢ psichomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Loading SRA and user-provided RNA-seq data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156449666-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156449666-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">psichomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GUI_tutorial.html">Visual interface tutorial</a>
    </li>
    <li>
      <a href="../articles/CLI_tutorial.html">Command-line interface (CLI) tutorial</a>
    </li>
    <li>
      <a href="../articles/custom_data.html">Loading SRA and user-provided RNA-seq data</a>
    </li>
    <li>
      <a href="../articles/AS_events_preparation.html">Preparing custom alternative splicing annotations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Loading SRA and user-provided RNA-seq data</h1>
                        <h4 class="author">Nuno Saraiva-Agostinho</h4>
            
            <h4 class="date">2020-02-13</h4>
      
      
      <div class="hidden name"><code>custom_data.Rmd</code></div>

    </div>

    
    
<hr>
<p><em>psichomics</em> is an interactive R package for integrative analyses of alternative splicing and gene expression based on <a href="https://tcga-data.nci.nih.gov/docs/publications/tcga">The Cancer Genome Atlas (TCGA)</a> (containing molecular data associated with 34 tumour types), the <a href="https://www.gtexportal.org">Genotype-Tissue Expression (GTEx)</a> project (containing data for multiple normal human tissues), <a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive</a> and user-provided data.</p>
<p>The following tutorial goes through the steps required to load custom RNA-seq data:</p>
<ol style="list-style-type: decimal">
<li>Retrieve FASTQ files and sample-associated information from the <a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (SRA)</a> (optional if you already have your own FASTQ samples);</li>
<li>Map RNA-seq reads from the FASTQ files against a genome of reference using <a href="https://github.com/alexdobin/STAR">STAR</a>, a splice-aware aligner;</li>
<li>Merge and prepare the output of such aligners to be correctly interpreted by psichomics;</li>
<li>Load data into psichomics.</li>
</ol>
<div id="download-data-from-sra-optional" class="section level1">
<h1 class="hasAnchor">
<a href="#download-data-from-sra-optional" class="anchor"></a>Download data from SRA (optional)</h1>
<p><a href="https://www.ncbi.nlm.nih.gov/sra">SRA</a> is a repository of biological sequence data that stores data from many published articles. <a href="https://www.ncbi.nlm.nih.gov/sra">SRA</a> data may be useful to answer pressing biological questions using publicly available data.</p>
<blockquote>
<p>The latest versions of psichomics support <strong>automatic downloading of SRA data</strong> from <a href="https://jhubiostatistics.shinyapps.io/recount/">recount2</a>, a resource of pre-processed data for thousands of SRA projects (including gene read counts, splice junction quantification and sample metadata). Check first if the project of your interest is not available through this resource, thus making it easier to analyse gene expression and alternative splicing for your samples of interest.</p>
</blockquote>
<p>Data from SRA can be downloaded using the <a href="https://ncbi.github.io/sra-tools/fastq-dump.html">fastq-dump</a> command from <a href="https://ncbi.github.io/sra-tools/">sra-tools</a>. For instance, to retrieve samples from the <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP126561">SRP126561</a> project, we could do the following:</p>
<pre><code># List SRA samples
samples=(SRR6368612 SRR6368613 SRR6368614 SRR6368615 SRR6368616 SRR6368617)

# Download each sample using fastq-dump
for each in ${samples}; do
    fastq-dump --gzip --split-3 ${each}
done</code></pre>
<p>Arguments used in the previous command:</p>
<ul>
<li>
<strong><code>--gzip</code></strong>: compress FASTQ files in the GZIP format</li>
<li>
<strong><code>--split-3</code></strong>: allows to output one or two FASTQ files for single-end or paired-end sequencing, respectively (a third FASTQ file may also be returned containing orphaned single-end reads obtained from paired-end sequencing data)</li>
</ul>
<p>Sample-associated data is also available from the <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP126561">Run Selector</a> page. Click <strong>RunInfo Table</strong> to download the whole metadata table for all samples.</p>
</div>
<div id="align-rna-seq-data-to-quantify-splice-junctions" class="section level1">
<h1 class="hasAnchor">
<a href="#align-rna-seq-data-to-quantify-splice-junctions" class="anchor"></a>Align RNA-seq data to quantify splice junctions</h1>
<p>The quantification of each alternative splicing event is based on the proportion of junction reads that support the inclusion isoform, known as percent spliced-in or PSI <span class="citation">(Wang <em>et al.</em>, 2008)</span>.</p>
<p>To estimate this value for each splicing event, both alternative splicing annotation and quantification of RNA-Seq reads aligning to splice junctions (junction quantification) are required. While alternative splicing annotation is provided by the package, junction quantification will need to be prepared from user-provided data by aligning the RNA-seq reads from FASTQ files to a genome of reference. As junction reads are required to quantify alternative splicing, a splice-aware aligner will be used. psichomics currently supports STAR output.</p>
<div id="star" class="section level2">
<h2 class="hasAnchor">
<a href="#star" class="anchor"></a>STAR</h2>
<p>Before aligning FASTQ samples against a genome of reference, a genome index needs to be prepared.</p>
<p>Start by downloading a FASTA file of the whole genome and a GTF file with annotated transcript. To run the following example, download the <a href="https://grch37.ensembl.org/info/data/ftp/">human FASTA and GTF files (hg19 assembly)</a>.</p>
<pre><code>mkdir hg19_STAR
STAR --runThreadN 4 \
     --runMode genomeGenerate \
     --genomeDir hg19_STAR \
     --genomeFastaFiles /path/to/hg19.fa \
     --sjdbGTFfile /path/to/hg19.gtf</code></pre>
<p>Arguments used in the previous command:</p>
<ul>
<li>
<strong><code>--runThreadN 4</code></strong>: run STAR in parallel using 4 threads</li>
<li>
<strong><code>--runMode genomeGenerate</code></strong>: generate the genome index</li>
<li>
<strong><code>--genomeDir</code></strong>: directory of the genome index directory (output)</li>
<li>
<strong><code>--genomeFastaFiles</code></strong>: path to genome file (FASTA format)</li>
<li>
<strong><code>--sjdbGTFfile</code></strong>: path to transcript annotation (GTF format)</li>
</ul>
<hr>
<p>After the genome index is generated, the sequences in the FASTQ files are aligned against the annotated gene and splice junctions from the previously prepared reference. The following commands make STAR output both gene and junction read counts into files ending in <code>ReadsPerGene.out.tab</code> and <code>SJ.out.tab</code>, respectively.</p>
<pre><code>align () {
    echo "Aligning ${1} using STAR..."
    STAR --runThreadN 16 \
         --genomeDir hg19_STAR \
         --quantMode GeneCounts \
         --readFilesCommand zcat \
         --outFileNamePrefix ${1} \
         --readFilesIn ${1}_1.fastq.gz ${1}_2.fastq.gz
}

for each in ${samples}; do
    align "${each}"
done</code></pre>
<p>Arguments used in the previous command:</p>
<ul>
<li>
<strong><code>--runThreadN 4</code></strong>: run STAR in parallel using 4 threads</li>
<li>
<strong><code>--genomeDir</code></strong>: directory of the genome index directory (input)</li>
<li>
<strong><code>--quantMode GeneCounts</code></strong>: count reads aligning per gene</li>
<li>
<strong><code>--readFilesCommand</code></strong>: command to extract compressed FASTQ files</li>
<li>
<strong><code>--outFileNamePrefix</code></strong>: output prefix</li>
<li>
<strong><code>--readFilesIn</code></strong>: FASTQ files to align</li>
</ul>
</div>
</div>
<div id="prepare-output-for-psichomics" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-output-for-psichomics" class="anchor"></a>Prepare output for psichomics</h1>
<p>To process the resulting data files, open an R console or RStudio and type:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Change working directory to where the STAR output is</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">"/path/to/aligned/output/"</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(psichomics)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw"><a href="../reference/prepareSRAmetadata.html">prepareGeneQuant</a></span>(</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">    <span class="st">"SRR6368612ReadsPerGene.out.tab"</span>, <span class="st">"SRR6368613ReadsPerGene.out.tab"</span>,</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">    <span class="st">"SRR6368614ReadsPerGene.out.tab"</span>, <span class="st">"SRR6368615ReadsPerGene.out.tab"</span>,</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">    <span class="st">"SRR6368616ReadsPerGene.out.tab"</span>, <span class="st">"SRR6368617ReadsPerGene.out.tab"</span>)</a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="kw"><a href="../reference/prepareSRAmetadata.html">prepareJunctionQuant</a></span>(<span class="st">"SRR6368612SJ.out.tab"</span>, <span class="st">"SRR6368613SJ.out.tab"</span>, </a>
<a class="sourceLine" id="cb4-10" data-line-number="10">                     <span class="st">"SRR6368614SJ.out.tab"</span>, <span class="st">"SRR6368615SJ.out.tab"</span>,</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">                     <span class="st">"SRR6368616SJ.out.tab"</span>, <span class="st">"SRR6368617SJ.out.tab"</span>)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="kw"><a href="../reference/prepareSRAmetadata.html">prepareSRAmetadata</a></span>(<span class="st">"SraRunTable.txt"</span>)</a></code></pre></div>
<div id="load-data-in-visual-interface" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data-in-visual-interface" class="anchor"></a>Load data in visual interface</h2>
<p>Start psichomics with the following commands in an R console or RStudio:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(psichomics)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw"><a href="../reference/psichomics.html">psichomics</a></span>()</a></code></pre></div>
<p>Then, click <strong>Load user files</strong>. Click the <strong>Folder input</strong> tab and select the appropriate folder where the psichomics-prepared data is stored. Finally, click <strong>Load files</strong> to automatically scan and load supported files in the selected folder.</p>
</div>
<div id="load-data-in-command-line-interface-cli" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data-in-command-line-interface-cli" class="anchor"></a>Load data in command-line interface (CLI)</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(psichomics)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">data          &lt;-<span class="st"> </span><span class="kw"><a href="../reference/loadLocalFiles.html">loadLocalFiles</a></span>(<span class="st">"/path/to/aligned/output/"</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">geneExpr      &lt;-<span class="st"> </span>data[[<span class="dv">1</span>]]<span class="op">$</span><span class="st">`</span><span class="dt">Gene expression</span><span class="st">`</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">junctionQuant &lt;-<span class="st"> </span>data[[<span class="dv">1</span>]]<span class="op">$</span><span class="st">`</span><span class="dt">Junction quantification</span><span class="st">`</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">sampleInfo    &lt;-<span class="st"> </span>data[[<span class="dv">1</span>]]<span class="op">$</span><span class="st">`</span><span class="dt">Sample metadata</span><span class="st">`</span></a></code></pre></div>
</div>
</div>
<div id="feedback" class="section level1">
<h1 class="hasAnchor">
<a href="#feedback" class="anchor"></a>Feedback</h1>
<p>All feedback on the program, documentation and associated material (including this tutorial) is welcome. Please send any comments and questions to:</p>
<blockquote>
<p>Nuno Saraiva-Agostinho (<a href="mailto:nunoagostinho@medicina.ulisboa.pt">nunoagostinho@medicina.ulisboa.pt</a>)</p>
<p><a href="http://imm.medicina.ulisboa.pt/group/distrans/">Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)</a></p>
</blockquote>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-wang2008">
<p>Wang,E.T. <em>et al.</em> (2008) Alternative isoform regulation in human tissue transcriptomes. <em>Nature</em>, <strong>456</strong>, 470â€“476.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#download-data-from-sra-optional">Download data from SRA (optional)</a></li>
      <li>
<a href="#align-rna-seq-data-to-quantify-splice-junctions">Align RNA-seq data to quantify splice junctions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#star">STAR</a></li>
      </ul>
</li>
      <li>
<a href="#prepare-output-for-psichomics">Prepare output for psichomics</a><ul class="nav nav-pills nav-stacked">
<li><a href="#load-data-in-visual-interface">Load data in visual interface</a></li>
      <li><a href="#load-data-in-command-line-interface-cli">Load data in command-line interface (CLI)</a></li>
      </ul>
</li>
      <li><a href="#feedback">Feedback</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Nuno Saraiva-Agostinho, Nuno LuÃ­s Barbosa-Morais.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
